lib/bulk/common.js

```diff
@@ -441,13 +444,519 @@ const BulkWriteError = function(error, result) {
+ */
+const BulkOperation = function(topology, collection, options, bulk) {
+  options = options == null ? {} : options;
+  // TODO Bring from driver information in isMaster
+  // Get the namespace for the write operations
+  var namespace = collection.collectionName;
+  // Used to mark operation as executed
+  var executed = false;
+
+  // Current item
+  var currentOp = null;
+
+  // Handle to the bson serializer, used to calculate running sizes
+  var bson = topology.bson;
+
+  // Set max byte size
+  var maxBatchSizeBytes =
+    topology.isMasterDoc && topology.isMasterDoc.maxBsonObjectSize
+      ? topology.isMasterDoc.maxBsonObjectSize
+      : 1024 * 1025 * 16;
+  var maxWriteBatchSize =
+    topology.isMasterDoc && topology.isMasterDoc.maxWriteBatchSize
+      ? topology.isMasterDoc.maxWriteBatchSize
+      : 1000;
+
+  // Get the write concern
+  var writeConcern = applyWriteConcern(shallowClone(options), { collection: collection }, options);
+  writeConcern = writeConcern.writeConcern;
+
+  // Get the promiseLibrary
+  var promiseLibrary = options.promiseLibrary || Promise;
+
+  // Final results
+  var bulkResult = {
+    ok: 1,
+    writeErrors: [],
+    writeConcernErrors: [],
+    insertedIds: [],
+    nInserted: 0,
+    nUpserted: 0,
+    nMatched: 0,
+    nModified: 0,
+    nRemoved: 0,
+    upserted: []
+  };
+
+  // Internal state
+  bulk.s = {
+    // Final result
+    bulkResult: bulkResult,
+    // Current batch state
+    currentBatch: null,
+    currentIndex: 0,
+    // ordered specific
+    currentBatchSize: 0,
+    currentBatchSizeBytes: 0,
+    // unordered specific
+    currentInsertBatch: null,
+    currentUpdateBatch: null,
+    currentRemoveBatch: null,
+    batches: [],
+    // Write concern
+    writeConcern: writeConcern,
+    // Max batch size options
+    maxBatchSizeBytes: maxBatchSizeBytes,
+    maxWriteBatchSize: maxWriteBatchSize,
+    // Namespace
+    namespace: namespace,
+    // BSON
+    bson: bson,
+    // Topology
+    topology: topology,
+    // Options
+    options: options,
+    // Current operation
+    currentOp: currentOp,
+    // Executed
+    executed: executed,
+    // Collection
+    collection: collection,
+    // Promise Library
+    promiseLibrary: promiseLibrary,
+    // Fundamental error
+    err: null,
+    // check keys
+    checkKeys: typeof options.checkKeys === 'boolean' ? options.checkKeys : true
+  };
+
+  // bypass Validation
+  if (options.bypassDocumentValidation === true) {
+    bulk.s.bypassDocumentValidation = true;
+  }
+};
+
+/**
+ * Add a single insert document to the bulk operation
+ *
+ * @param {object} document the document to insert
+ * @param {class} bulk either OrderedBulkOperation or UnorderdBulkOperation
+ * @throws {MongoError}
+ * @return {OrderedBulkOperation or UnorderedBulkOperation}
+ */
+const Insert = function(document, bulk) {
+  if (bulk.s.collection.s.db.options.forceServerObjectId !== true && document._id == null)
```

Comment:
Did this line exist in the original? I am kind of nervous about an access chain this long.

Reply:
This line did exist in the original. Should I change it?

Reply:
No, leave it as is for now. :(